File: app/src/main/java/com/mvcoding/financius/ui/transaction/TransactionActivity.java
/*
 * Copyright (C) 2015 Mantas Varnagiris.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package com.mvcoding.financius.ui.transaction;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioGroup;

import com.mvcoding.financius.R;
import com.mvcoding.financius.core.model.TransactionState;
import com.mvcoding.financius.core.model.TransactionType;
import com.mvcoding.financius.data.model.Place;
import com.mvcoding.financius.data.model.Tag;
import com.mvcoding.financius.data.model.Transaction;
import com.mvcoding.financius.ui.ActivityComponent;
import com.mvcoding.financius.ui.ActivityStarter;
import com.mvcoding.financius.ui.BaseActivity;
import com.mvcoding.financius.ui.DateDialogFragment;
import com.mvcoding.financius.ui.TimeDialogFragment;
import com.mvcoding.financius.ui.calculator.CalculatorActivity;
import com.mvcoding.financius.util.date.DateFormatter;
import com.mvcoding.financius.util.rx.Event;
import com.mvcoding.financius.util.rx.RxBus;

import org.joda.time.DateTime;

import java.math.BigDecimal;
import java.util.Set;

import javax.inject.Inject;

import butterknife.Bind;
import rx.Observable;
import rx.android.view.ViewObservable;
import rx.android.widget.WidgetObservable;
import rx.subjects.PublishSubject;

public class TransactionActivity extends BaseActivity<TransactionPresenter.View, TransactionComponent> implements TransactionPresenter.View {
    private static final String EXTRA_TRANSACTION = "EXTRA_TRANSACTION";

    private static final String RESULT_EXTRA_TRANSACTION = "RESULT_EXTRA_TRANSACTION";

    private static final int REQUEST_DATE = 1;
    private static final int REQUEST_TIME = 2;
    private static final int REQUEST_AMOUNT = 3;

    private static final PublishSubject<TransactionType> transactionTypeSubject = PublishSubject.create();
    private static final PublishSubject<TransactionState> transactionStateSubject = PublishSubject.create();
    private static final PublishSubject<Long> dateSubject = PublishSubject.create();
    private static final PublishSubject<BigDecimal> amountSubject = PublishSubject.create();
    private static final PublishSubject<String> currencySubject = PublishSubject.create();
    private static final PublishSubject<Place> placeSubject = PublishSubject.create();
    private static final PublishSubject<Set<Tag>> tagsSubject = PublishSubject.create();

    @Bind(R.id.transactionTypeRadioGroup) RadioGroup transactionTypeRadioGroup;
    @Bind(R.id.transactionStateRadioGroup) RadioGroup transactionStateRadioGroup;
    @Bind(R.id.dateButton) Button dateButton;
    @Bind(R.id.timeButton) Button timeButton;
    @Bind(R.id.amountButton) Button amountButton;
    @Bind(R.id.currencyButton) Button currencyButton;
    @Bind(R.id.placeButton) Button placeButton;
    @Bind(R.id.tagsButton) Button tagsButton;
    @Bind(R.id.noteEditText) EditText noteEditText;
    @Bind(R.id.saveButton) Button saveButton;

    @Inject TransactionPresenter presenter;
    @Inject RxBus rxBus;

    private boolean ignoreChanges = false;
    private Transaction transaction;

    public static void startForResult(@NonNull Context context, int requestCode, @NonNull Transaction transaction) {
        ActivityStarter.with(context, TransactionActivity.class).extra(EXTRA_TRANSACTION, transaction).startForResult(requestCode);
    }

    @Override protected int getLayoutId() {
        return R.layout.activity_transaction;
    }

    @Override protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        transactionTypeRadioGroup.setOnCheckedChangeListener((group, checkedId) -> transactionTypeSubject.onNext(checkedId == R.id.transactionTypeExpenseRadioButton ? TransactionType.Expense : TransactionType.Income));
        transactionStateRadioGroup.setOnCheckedChangeListener((group, checkedId) -> transactionStateSubject.onNext(checkedId == R.id.transactionStateConfirmedRadioButton ? TransactionState.Confirmed : TransactionState.Pending));
        amountButton.setOnClickListener(v -> CalculatorActivity.startForResult(this, REQUEST_AMOUNT, transaction.getAmount()));
    }

    @Override protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case REQUEST_AMOUNT:
                if (resultCode == RESULT_OK) {
                    amountSubject.onNext(CalculatorActivity.getResultNumber(data));
                }
                break;
        }
    }

    @NonNull @Override protected TransactionComponent createComponent(@NonNull ActivityComponent component) {
        final Transaction transaction = getIntent().getParcelableExtra(EXTRA_TRANSACTION);
        return component.plus(new TransactionModule(transaction));
    }

    @Override protected void inject(@NonNull TransactionComponent component) {
        component.inject(this);
    }

    @NonNull @Override protected TransactionPresenter getPresenter() {
        return presenter;
    }

    @NonNull @Override protected TransactionPresenter.View getPresenterView() {
        return this;
    }

    @NonNull @Override public Observable<TransactionType> onTransactionTypeChanged() {
        return transactionTypeSubject.filter(transactionType -> !ignoreChanges);
    }

    @NonNull @Override public Observable<TransactionState> onTransactionStateChanged() {
        return transactionStateSubject.filter(transactionState -> !ignoreChanges);
    }

    @NonNull @Override public Observable<Long> onDateChanged() {
        final Observable<Long> dateObservable = rxBus.observe(DateDialogFragment.DateDialogResult.class)
                .mergeWith(ViewObservable.clicks(dateButton)
                                   .flatMap(onClickEvent -> DateDialogFragment.show(getSupportFragmentManager(), REQUEST_DATE, rxBus, transaction
                                           .getDate())))
                .map(dateDialogResult -> {
                    final DateTime dateTime = new DateTime(transaction.getDate());
                    return new DateTime(dateDialogResult.getYear(), dateDialogResult.getMonthOfYear(), dateDialogResult.getDayOfMonth(), dateTime
                            .getHourOfDay(), dateTime.getMinuteOfHour()).getMillis();
                });

        final Observable<Long> timeObservable = rxBus.observe(TimeDialogFragment.TimeDialogResult.class)
                .mergeWith(ViewObservable.clicks(timeButton)
                                   .flatMap(onClickEvent -> TimeDialogFragment.show(getSupportFragmentManager(), REQUEST_TIME, rxBus, transaction
                                           .getDate())))
                .map(timeDialogResult -> {
                    final DateTime dateTime = new DateTime(transaction.getDate());
                    return new DateTime(dateTime.getYear(), dateTime.getMonthOfYear(), dateTime.getDayOfMonth(), timeDialogResult.getHourOfDay(), timeDialogResult
                            .getMinuteOfHour()).getMillis();
                });

        return Observable.merge(dateObservable, timeObservable);
    }

    @NonNull @Override public Observable<BigDecimal> onAmountChanged() {
        return amountSubject;
    }

    @NonNull @Override public Observable<String> onCurrencyChanged() {
        return currencySubject;
    }

    @NonNull @Override public Observable<Place> onPlaceChanged() {
        return placeSubject;
    }

    @NonNull @Override public Observable<Set<Tag>> onTagsChanged() {
        return tagsSubject;
    }

    @NonNull @Override public Observable<String> onNoteChanged() {
        return WidgetObservable.text(noteEditText).map(onTextChangeEvent -> onTextChangeEvent.text().toString());
    }

    @NonNull @Override public Observable<Event> onSave() {
        return ViewObservable.clicks(saveButton).map(onClickEvent -> new Event());
    }

    @Override public void showTransaction(@NonNull Transaction transaction) {
        this.transaction = transaction;
        ignoreChanges = true;

        transactionTypeRadioGroup.check(transaction.getTransactionType() == TransactionType.Expense ? R.id.transactionTypeExpenseRadioButton : R.id.transactionTypeIncomeRadioButton);
        transactionStateRadioGroup.check(transaction.getTransactionState() == TransactionState.Confirmed ? R.id.transactionStateConfirmedRadioButton : R.id.transactionStatePendingRadioButton);
        dateButton.setText(DateFormatter.date(this, transaction.getDate()));
        timeButton.setText(DateFormatter.time(this, transaction.getDate()));
        amountButton.setText(transaction.getAmount().toString());

        ignoreChanges = false;
    }

    @Override public void startResult(@NonNull Transaction transaction) {
        final Intent data = new Intent();
        data.putExtra(RESULT_EXTRA_TRANSACTION, transaction);
        setResult(RESULT_OK, data);
    }
}

File: app/src/main/kotlin/com/mvcoding/expensius/feature/transaction/TransactionActivity.kt
/*
 * Copyright (C) 2018 Mantas Varnagiris.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package com.mvcoding.expensius.feature.transaction

//class TransactionActivity : BaseActivity(), TransactionPresenter.View {
//    companion object {
//        private const val EXTRA_TRANSACTION = "EXTRA_TRANSACTION"
//
//        private const val REQUEST_AMOUNT = 1
//        private const val REQUEST_DATE = 2
//        private const val REQUEST_EXCHANGE_RATE = 3
//
//        fun start(context: Context, transaction: Transaction) {
//            ActivityStarter(context, TransactionActivity::class)
//                    .extra(EXTRA_TRANSACTION, transaction)
//                    .start()
//        }
//    }
//
//    private val presenter by lazy { provideTransactionPresenter(intent.getSerializableExtra(EXTRA_TRANSACTION) as Transaction) }
//    private val moneyFormatter by lazy { provideMoneyFormatter() }
//    private val dateFormatter by lazy { provideDateFormatter() }
//    private val amountSubject by lazy { PublishSubject.create<BigDecimal>() }
//    private val exchangeRateSubject by lazy { PublishSubject.create<BigDecimal>() }
//    private val rxBus by lazy { provideRxBus() }
//
//    private var isArchiveToggleVisible = true
//    private var archiveToggleTitle: String? = null
//    private var timestamp = 0L
//    private var amount = ZERO
//
//    override fun onCreate(savedInstanceState: Bundle?) {
//        super.onCreate(savedInstanceState)
//        setContentView(R.layout.activity_transaction)
//
//        amountTextView.clicks().subscribe { CalculatorActivity.startWithInitialNumberForResult(this, REQUEST_AMOUNT, amount) }
//        dateButton.clicks().subscribe { DateDialogFragment.show(supportFragmentManager, REQUEST_DATE, rxBus, timestamp) }
//
//        presenter.attach(this)
//    }
//
//    override fun onDestroy() {
//        presenter.detach(this)
//        super.onDestroy()
//    }
//
//    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
//        super.onActivityResult(requestCode, resultCode, data)
//        if (resultCode != RESULT_OK) return
//
//        when (requestCode) {
//            REQUEST_AMOUNT -> amountSubject.onNext(CalculatorActivity.resultExtraAmount(data!!))
//            REQUEST_EXCHANGE_RATE -> exchangeRateSubject.onNext(CalculatorActivity.resultExtraAmount(data!!))
//        }
//    }
//
//    override fun onCreateOptionsMenu(menu: Menu): Boolean {
//        super.onCreateOptionsMenu(menu)
//        menuInflater.inflate(R.menu.tag, menu)
//        val menuItem = menu.findItem(R.id.action_archive)
//        menuItem.isVisible = isArchiveToggleVisible
//        menuItem.title = archiveToggleTitle
//        return true
//    }
//
//    override fun transactionStateChanges(): Observable<TransactionState> = transactionStateCheckBox.checkedChanges()
//            .map { if (it) CONFIRMED else PENDING }
//            .distinctUntilChanged()
//
//    override fun transactionTypeChanges(): Observable<TransactionType> = transactionTypeFloatingActionButton.clicks()
//            .map { if (transactionTypeFloatingActionButton.isSelected) EXPENSE else INCOME }
//            .distinctUntilChanged()
//
//    override fun timestampChanges(): Observable<Long> = rxBus.observe<DateDialogFragment.DateDialogResult>()
//            .map { DateTime(timestamp).withYear(it.year).withMonthOfYear(it.monthOfYear).withDayOfMonth(it.dayOfMonth).millis }
//
//    override fun exchangeRateChanges(): Observable<BigDecimal> = exchangeRateSubject
//    override fun amountChanges(): Observable<BigDecimal> = amountSubject
//    override fun tagsChanges(): Observable<Set<Tag>> = quickTagsView.selectedTagsChanges()
//    override fun noteChanges(): Observable<String> = noteEditText.textChanges().map { it.toString() }.distinctUntilChanged()
//    override fun archiveToggles(): Observable<Unit> = toolbar.itemClicks().filter { it.itemId == R.id.action_archive }.map { Unit }
//    override fun saveRequests(): Observable<Unit> = saveButton.clicks()
//    override fun currencyChangeRequests(): Observable<Unit> = currencyButton.clicks()
//
//    override fun currencyChanges(currencies: List<Currency>): Observable<Currency> = Observable.create {
//        val displayCurrencies = currencies.map { it.displayName() }
//        val itemHeight = getDimensionFromTheme(this, R.attr.actionBarSize)
//        val keyline = resources.getDimensionPixelSize(R.dimen.keyline)
//        val keylineHalf = resources.getDimensionPixelOffset(R.dimen.keyline_half)
//        val popupWindow = ListPopupWindow(this)
//        popupWindow.anchorView = currencyButton
//        popupWindow.setAdapter(ArrayAdapter<String>(this, R.layout.item_view_currency, R.id.currencyCodeTextView, displayCurrencies))
//        popupWindow.setOnItemClickListener { _, _, position, _ -> it.onNext(currencies[position]); popupWindow.dismiss() }
//        popupWindow.setOnDismissListener { it.onComplete() }
//        popupWindow.width = contentView.width - keyline
//        popupWindow.height = Math.min(contentView.height - buttonBarView.height - currencyButton.height - keylineHalf, itemHeight * 7)
//        popupWindow.isModal = true
//        popupWindow.horizontalOffset = keylineHalf
//        popupWindow.show()
//    }
//
//    override fun showArchiveEnabled(archiveEnabled: Boolean) {
//        isArchiveToggleVisible = archiveEnabled
//        toolbar.menu.findItem(R.id.action_archive)?.isVisible = archiveEnabled
//    }
//
//    override fun showModelState(modelState: ModelState) {
//        archiveToggleTitle = resources.getString(if (modelState == ModelState.NONE) R.string.archive else R.string.restore)
//        toolbar.menu.findItem(R.id.action_archive)?.title = archiveToggleTitle
//    }
//
//    override fun showTimestamp(timestamp: Timestamp) {
//        this.timestamp = timestamp.millis
//        dateButton.text = dateFormatter.formatDateRelativeToToday(timestamp.millis)
//    }
//
//    override fun showMoney(money: Money) {
//        amount = money.amount
//        amountTextView.text = moneyFormatter.format(money)
//        currencyButton.text = money.currency.displayName()
//    }
//
//    override fun showTransactionState(transactionState: TransactionState): Unit = with(transactionStateCheckBox) { isChecked = transactionState == CONFIRMED }
//    override fun showTransactionType(transactionType: TransactionType): Unit = with(transactionTypeFloatingActionButton) { isSelected = transactionType == INCOME }
//    override fun showTags(tags: Set<Tag>): Unit = quickTagsView.showSelectedTags(tags)
//    override fun showNote(note: Note): Unit = noteEditText.setTextIfChanged(note.text)
//    override fun displayResult(): Unit = finish()
//}